iothub_client/src/uamqp_messaging.c

```diff
@@ -479,30 +539,87 @@ static int readPropertiesFromuAMQPMessage(IOTHUB_MESSAGE_HANDLE iothub_message_h
-                LogError("Failed to get value of uAMQP message 'correlation-id' property (%d).", api_call_result);
-                return_value = __FAILURE__;
-            }
-            // Codes_SRS_UAMQP_MESSAGING_09_024: [The correlation-id property shall be set on the IOTHUB_MESSAGE_HANDLE by calling IoTHubMessage_SetCorrelationId, passing the value read from the uAMQP message.]
-            else if (IoTHubMessage_SetCorrelationId(iothub_message_handle, uamqp_message_property_value) != IOTHUB_MESSAGE_OK)
+            // Codes_SRS_UAMQP_MESSAGING_09_020: [The type of the correlation-id property value shall be obtained using amqpvalue_get_type().]
+            AMQP_TYPE value_type = amqpvalue_get_type(uamqp_message_property);
+
+            // Codes_SRS_UAMQP_MESSAGING_09_021: [If the type of the correlation-id property value is AMQP_TYPE_NULL, IoTHubMessage_CreateFromuAMQPMessage() shall skip processing the correlation-id (as it is optional) and continue normally.] 
+            if (value_type != AMQP_TYPE_NULL)
             {
-                // Codes_SRS_UAMQP_MESSAGING_09_025: [If IoTHubMessage_SetCorrelationId fails, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.]
-                LogError("Failed to set IOTHUB_MESSAGE_HANDLE 'correlation-id' property.");
-                return_value = __FAILURE__;
+                char* string_value = NULL;
+                char string_buffer[40];
+
+                if (value_type == AMQP_TYPE_STRING)
+                {
+                    // Codes_SRS_UAMQP_MESSAGING_09_022: [The correlation-id value shall be retrieved from the AMQP_VALUE as char sequence] 
+                    if (amqpvalue_get_string(uamqp_message_property, &string_value) != 0)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_023: [If correlation-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get value of uAMQP message 'correlation-id' property (string)");
+                        result = __FAILURE__;
+                    }
+                }
+                else if (value_type == AMQP_TYPE_ULONG)
+                {
+                    uint64_t ulong_value;
+
+                    // Codes_SRS_UAMQP_MESSAGING_09_022: [The correlation-id value shall be retrieved from the AMQP_VALUE as char sequence] 
+                    if (amqpvalue_get_ulong(uamqp_message_property, &ulong_value) != 0)
+                    {
+                        // Codes_SRS_UAMQP_MESSAGING_09_023: [If correlation-id fails to be obtained, message_create_IoTHubMessage_from_uamqp_message() shall fail and return immediately.] 
+                        LogError("Failed to get value of uAMQP message 'correlation-id' property (ulong)");
```

Comment:
>ulong [](start = 98, length = 5)

uint64_t

Reply:
I'll keep as is since it is mentioned as amqpvalue_get_ulong